src/utils/createComponent.ts

```diff
@@ -45,10 +41,30 @@ type Options<I extends HTMLElement, E extends EventNames = {}> = Readonly<{
   tagName: string;
 }>;
 
+// A map of expected event listener types based on EventNames.
+type EventListeners<R extends EventNames> = {
+  [K in keyof R]?: R[K] extends EventName ? (e: R[K]['__eventType']) => void : (e: Event) => void;
+};
+
+// Props derived from custom element class. Currently has limitations of making
+// all properties optional and also surfaces life cycle methods in autocomplete.
+// TODO: LoginOverlay has "autofocus" property, so we add it back manually.
+type ElementProps<I> = Partial<Omit<I, keyof HTMLElement>> & { autofocus?: boolean };
+
+// Acceptable props to the React component.
+type ComponentProps<I, E extends EventNames = {}> = Omit<
+  React.HTMLAttributes<I>,
+  // Prefer type of provided event handler props or those on element over
+  // built-in HTMLAttributes
+  keyof E | keyof ElementProps<I>
+> &
+  EventListeners<E> &
+  ElementProps<I>;
+
```

Comment:
We could possibly avoid copying the types with something like:
```ts
type Component<I extends HTMLElement, E extends EventNames = {}> = ReturnType<typeof _createComponent<I, E>>;
type ComponentProps<I extends HTMLElement, E extends EventNames = {}> = Parameters<Component<I, E>>[0];
```

Reply:
Tried to use this code and it results in the following error:

```
[dts] src/generated/ComboBox.ts(20,25): error TS2590: Expression produces a union type that is too complex to represent.
```

Reply:
Seems that with `React.ComponentProps` it doesn't happen though. Not blocking the PR for this.